### ISSUE TITLE
Optimizer (Adam) does not propagate hyperparameter update on first update

### ISSUE THREAD

**Issue by svobora** [tensorflow/tensorflow#31287]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): no
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04 64b
- TensorFlow version (use command below): 2.0.0b1
- Python version: 3.6

**Describe the current behavior**
Using tf.keras.optimizers.Adam and updating hyperparameters results in hyperparameters not updating on the first update.

**Describe the expected behavior**
Hyperparameters DO update on the first run, ie they are initialized to tf.Variables at constructor of the optimizer, not on first call.

**Code to reproduce the issue**
`adam = tf.keras.optimizers.Adam()`
`adam.learning_rate = 0.00001   # this does not update learning_rate, but changes python-typed ` `hyperparameters in "_hyper" dictionary to tf.Variables`
`adam.learning_rate = 0.00001  # this finally updates the tf.Variable `

Calling adam._hyper before update yields:

`{'learning_rate': 0.001, 'decay': 0.0, 'beta_1': 0.9, 'beta_2': 0.999}`

Calling adam._hyper after first update yields:

`{'learning_rate': <tf.Variable 'learning_rate:0' shape=() dtype=float32, numpy=0.001>, 'decay': <tf.Variable 'decay:0' shape=() dtype=float32, numpy=0.0>, 'beta_1': <tf.Variable 'beta_1:0' shape=() dtype=float32, numpy=0.9>, 'beta_2': <tf.Variable 'beta_2:0' shape=() dtype=float32, numpy=0.999>}`

**Comment by svobora** [comment]
Well actually to get it working, first a getter must be called for `adam.learning_rate` and only then the python-typed variables are replaced with tf.Variable variables and one can use the setter.

**Comment by gadagashwini** [comment]
@svobora, In order to expedite the trouble-shooting process, please provide a complete code to reproduce the issue reported here. Thanks!

**Comment by svobora** [comment]
Hello, I will provide a simple and complete, yet a little different example. I believe the actual problem is with the tf.train.Checkpoint. Probably the variables are overwritten by the restored values when adam._hyper is converted from plaintext format to tf.Variable format.

In the provided example, the optimizer.learning_rate is not properly stored. Even though it is a different problem, I believe they are connected. I am not yet able to easily reproduce the first problem, because it appears after ckpt.restore() call (I was able to reproduce it with minimal example), but it depends on the manager.save() (I was not yet able to reproduct with minimal example).


```
import tensorflow as tf

def save():
    optimizer = tf.keras.optimizers.Adam()
    ckpt = tf.train.Checkpoint(optimizer=optimizer)
    manager = tf.train.CheckpointManager(ckpt, "tmp", max_to_keep=3)
    _ = optimizer.learning_rate  # comment this line and "load()" will print (0.001)
    optimizer.learning_rate = 0.0003
    manager.save()

def load():
    pretrained_model = "tmp/ckpt-1"
    optimizer = tf.keras.optimizers.Adam()
    ckpt = tf.train.Checkpoint(optimizer=optimizer)
    ckpt.restore(str(pretrained_model))
    print("Learning rate for epoch {} set to {}.".format(0, optimizer.learning_rate))

if __name__ == '__main__':
    save()
    load()
```

Comment the suggested line to see a bug.

**Comment by gadagashwini** [comment]
@svobora Thanks for providing the reproducible code.
I am able to reproduce the issue with Tenosrflow 2.0.0.beta1 on Colab. Please see the colab [gist](https://colab.research.google.com/drive/1LmQbwnFbOcEF9g1XZvWWkqQjcgP-Dck_). Thanks!

**Comment by ymodak** [comment]
I think this issue surfaces a discrepancy in ```Save and Restore``` method and is not necessarily a bug on  ```tf.keras.optimizers.Adam``` side.
Can you please try setting ```learning_rate``` arg for the ```tf.keras.optimizers.Adam()``` instead?

**Comment by svobora** [comment]
I believe the main issue is that `tf.keras.optimizers.Adam()` returns not fully initialized object. What is the reason not to construct the object with `tf.Variable` hyperparameters in the first place? Why is the object constructed with hyperparameters in the form `{'learning_rate': 0.001, 'decay': 0.0, 'beta_1': 0.9, 'beta_2': 0.999}` and only on first getter call converted to `tf.Variable`?

**Comment by tanzhenyu** [comment]
Thanks for the report!
Variable creation is deferred so that Estimator use case is supported. (you can't create a variable outside of Estimator graph)
The workaround here is just to call opt._create_hypers() so that things will get initialized correctly before calling Checkpoint, but I'm concerned this is not a good solution.

**Comment by tanzhenyu** [comment]
@allenlavoie from the minimal example here, do you think it's better if we create variables eagerly, since checkpoint needs to capture variables at creation time?

**Comment by allenlavoie** [comment]
How about the setter also creates a Variable? So if you pass the constructor argument you don't get a variable (since presumably you'll always pass that same value, so no need to restore it from a checkpoint) but if you assign to the attribute you always get a variable.

**Comment by tanzhenyu** [comment]
You can simply initialize the optimizer with the right learning_rate. This seems a niche case to support. Closing it for now.

**Comment by svobora** [comment]
_You can simply initialize the optimizer with the right learning_rate._

Really? But maybe I should be aware first that save function called on an object doesn't save anything to know that I have to set it manually. The issue is in unexpected behavior of the Saver, which depends on background creation of some other object no user knows about. Like the @allenlavoie  said, either create on assignment, or ban assignment. Do not allow to assign to preliminary objects.

**Comment by svobora** [comment]
What I ended up doing is calling a getter on the object, which as opposed to setter, actually creates the Variable.

`_ = optimizer.learning_rate`

**Comment by tanzhenyu** [comment]
> What I ended up doing is calling a getter on the object, which as opposed to setter, actually creates the Variable.
> 
> `_ = optimizer.learning_rate`

This is only a workaround, b/c there are quite some other hyper parameters. I'm not sure what your use case is, why would you use an optimizer without minimizing anything?
On the other hand, we're moving forward with having Adam(var_list) or something like that, which should be able to fix this problem

**Comment by svobora** [comment]
_why would you use an optimizer without minimizing anything?_

How did you infer that I do not minimize anything? I posted a toy example to reproduce the issue, not my company's code, which was affected and I had to find the issue and fix it.

_On the other hand, we're moving forward with having Adam(var_list) or something like that, which should be able to fix this problem_

Great! Hopefully things will improve over time. Thx for reply.

**Comment by tanzhenyu** [comment]
> _why would you use an optimizer without minimizing anything?_
> 
> How did you infer that I do not minimize anything? I posted a toy example to reproduce the issue, not my company's code, which was affected and I had to find the issue and fix it.
If you minimize a list of variables, it will create the hyper parameters as tf.Variables after it, so the saving and loading would work.
> 
> _On the other hand, we're moving forward with having Adam(var_list) or something like that, which should be able to fix this problem_
> 
> Great! Hopefully things will improve over time. Thx for reply.

**Comment by svobora** [comment]
_If you minimize a list of variables, it will create the hyper parameters as tf.Variables after it, so the saving and loading would work._

Could you point me to a documentation that specifies you cannot call Saver before minimizing a set of variables?

**Comment by tanzhenyu** [comment]
> _If you minimize a list of variables, it will create the hyper parameters as tf.Variables after it, so the saving and loading would work._
> 
> Could you point me to a documentation that specifies you cannot call Saver before minimizing a set of variables?

What I mean is it will serve as a workaround for you before this is fixed in nightly or 2.3

**Comment by svobora** [comment]
OK, thanks!
